src/App.tsx

```diff
@@ -1,24 +1,207 @@
+          setError(null);
+        }, 3000);
+      });
+  }, []);
+
+  const deleteErrorMessage = () => {
+    setError(null);
+  };
+
+  const deleteErrorMessageAfterDelay = (delay: number) => {
+    setTimeout(() => {
+      setError(null);
+    }, delay);
+  };
+
+  const getVisibleTodos = () => {
+    const needsToFilter
+      = todoSelector === TodoSelector.ACTIVE
+      || todoSelector === TodoSelector.COMPLETED;
+
+    if (!needsToFilter) {
+      return todos;
+    }
+
+    return todos.filter((todo) => {
+      switch (todoSelector) {
+        case TodoSelector.ACTIVE:
+          return !todo.completed;
+        case TodoSelector.COMPLETED:
+          return todo.completed;
+        default:
+          return true;
+      }
+    });
+  };
+
+  const visibleTodos = useMemo(getVisibleTodos, [todos, todoSelector]);
+
+  const hasCompletedTodos = todos.some((todo) => todo.completed);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const leftTodosCount = todos.reduce((acc, todo) => {
+    return !todo.completed ? acc + 1 : acc;
+  }, 0);
+
+  const handleTodoSelection = (event: React.MouseEvent<HTMLAnchorElement>) => {
+    setTodoSelector(event.currentTarget.textContent);
+  };
+
+  const handleTodoInputChanging = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    const { value } = event.target;
+
+    setInputValue(value);
+  };
+
+  const handleAddTodo = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!inputValue.trim()) {
+      setError(new Error("Title can't be empty"));
+      deleteErrorMessageAfterDelay(3000);
+
+      return;
+    }
+
+    const newTodo: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title: inputValue,
+      completed: false,
+    };
+
+    postTodo(newTodo)
+      .then((response) => {
+        setTodos((t) => [...t, response]);
+        setTempTodo(null);
+        setInputValue('');
+      })
+      .catch(() => {
+        setError(new Error('Unable to add a todo'));
+        deleteErrorMessageAfterDelay(3000);
+      });
+
+    setTempTodo(newTodo);
+  };
+
+  const handleDeleteTodo = (todoId: number) => () => {
+    setIsDeleting(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos((prevTodos) => prevTodos.filter((todo) => todo.id !== todoId));
+        setIsDeleting(false);
+      })
+      .catch(() => {
+        setError(new Error('Unable to delete a todo'));
+        deleteErrorMessageAfterDelay(3000);
+      });
+  };
+
+  const handleCompletedTodoDeleting = () => {
+    todos
+      .filter((todo) => todo.completed)
+      .forEach((todo) => {
+        setIsDeleting(true);
+
+        deleteTodo(todo.id)
+          .then(() => {
+            setTodos((prevTodos) => prevTodos.filter((t) => t.id !== todo.id));
+            setIsDeleting(false);
+          })
+          .catch(() => {
+            setError(new Error('Unable to delete a todo'));
+            deleteErrorMessageAfterDelay(3000);
+          });
```

Comment:
Do not repeat your self
```suggestion
        handleDeleteTodo(todo.id);
```